### PR TITLE
runtime(uci): add support

### PR DESCRIPTION
**Pull request by cmcaine**
Updated version of #14385

This version adds content detection for files that have a line starting with "c \S", "config \S" or "p \S" or "package \S" in the first 3 lines.

The detection only checks the second line if the first line looked like it was blank or a comment. Similarly for line 3.

### CODE REVIEW

File: runtime/ftplugin/uci.vim [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+let b:did_ftplugin = 1
+
+" UCI files are indented with tabs.
+setl noexpandtab
+setl shiftwidth=0
+setl softtabstop=0
+
+setl commentstring=#\ %s
+
+let b:undo_ftplugin = "setlocal et< cms<"
```
Comment (chrisbra) [inline runtime/ftplugin/uci.vim]: missing shiftwidth and softatbstop

File: runtime/autoload/dist/ft.vim [hunk 2]
```diff
@@ -1298,5 +1298,21 @@ export def FTvba()
...
+  # Match a line that is either all blank or blank followed by a comment
+  const comment_or_blank = '^\s*\(#.*\)\?$'
+
+  # Return true iff the file has a config or package statement near the
+  # top of the file and all preceding lines were comments or blank.
+  return getline(1) =~# config_or_package_statement
+  \   || getline(1) =~# comment_or_blank
+  \      && (   getline(2) =~# config_or_package_statement
+  \          || getline(2) =~# comment_or_blank
+  \             && getline(3) =~# config_or_package_statement
```
Comment (chrisbra) [inline runtime/autoload/dist/ft.vim:1313]: is this okay to only check the first 3 lines? The logic is a bit ambigious: `expr || expr && expr`.

Comment (cmcaine) [reply runtime/autoload/dist/ft.vim:1313]: What's ambiguous? && has higher precedence than ||, so the meaning of the code is well defined.

> is this okay to only check the first 3 lines?

I don't understand this question either, sorry. Can you rephrase it, please?

File: runtime/ftplugin/uci.vim (hunk 1 above)
Comment (cmcaine) [reply runtime/ftplugin/uci.vim]:
```suggestion
let b:undo_ftplugin = "setlocal et< cms< sts< sw<"
```

File: runtime/autoload/dist/ft.vim (hunk 2 above)
Comment (chrisbra) [reply runtime/autoload/dist/ft.vim:1313]: > What's ambiguous? && has higher precedence than ||, so the meaning of the code is well defined.

It might be well-defined, but I prefer to be explict, so I don't have to remember it

Comment (chrisbra) [reply runtime/autoload/dist/ft.vim:1313]: > I don't understand this question either, sorry. Can you rephrase it, please?

Is it enough to only check the first 3 lines? Can't there be 20 line comments first, and only then a `config_or_package_statement`?

Comment (cmcaine) [reply runtime/autoload/dist/ft.vim:1313]: && has higher precedence in every programming language I remember using it in, but sure, I don't really care if there are extra parentheses in there or not. Are you requesting that I add them?

The heuristic can be wrong if it checks 0 or 1000 lines. 3 lines was chosen arbitrarily. E.g. it might be better if a blank file with path `sshfs/router1/etc/config/firewall` gets UCI syntax highlighting, but the current detection code would not assign it the uci filetype because it is trying to avoid false positives at the expense of having more false negatives.

Honestly, I regret mentioning that the path matches perhaps too many files 🥲

Comment (chrisbra) [reply runtime/autoload/dist/ft.vim:1313]: alright fine.

File: src/testdir/test_filetype.vim [hunk 3]
```diff
@@ -764,6 +764,7 @@ def s:GetFilenameChecks(): dict<list<string>>
     typespec: ['file.tsp'],
     ungrammar: ['file.ungram'],
     uc: ['file.uc'],
+    uci: ['any/etc/config/file'],
```
Comment (chrisbra) [inline src/testdir/test_filetype.vim]: that Test fails. I think you need to create a file with specific content to make this work, see e.g. Test_def_file()

Comment (cmcaine) [reply src/testdir/test_filetype.vim]: I think this is now resolved.

**Comment by chrisbra** [comment]
Thanks, let me include it now

**Comment by cmcaine** [comment]
Thanks!

**Comment by Shane-XB-Qian** [comment]
> && has higher precedence in every programming language I remember using it in, but sure, I don't really care if there are extra parentheses in there or not.

@cmcaine unfortunately it is not always true if there was a bug :-) you check there is a bug seems last month tho that's for vim9.
just FYI, have fun anyway.

--
shane.xb.qian

**Comment by cmcaine** [comment]
Lol, well, hopefully somebody has added a regression test for that now :)
